libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1065,6 +1065,53 @@ string YulUtilFunctions::clearStorageArrayFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::clearStorageStructFunction(StructType const& _type)
+{
+	solAssert(_type.location() == DataLocation::Storage, "");
+
+	string functionName = "clear_struct_storage_" + _type.identifier();
+
+	return m_functionCollector.createFunction(functionName, [&] {
+		auto const& structType = dynamic_cast<StructType const&>(_type);
+		MemberList::MemberMap structMembers = structType.nativeMembers(nullptr);
+		vector<map<string, string>> memberSetValues;
+
+		for (auto member: structMembers)
+			if (member.type->isDynamicallyEncoded())
```

Comment:
I did say we need this for dynamically encoded types, but I think that wasn't entirely true, sorry about that :-).
The truth is a bit more annoying I'm afraid and will complicate this a bit further...

Statically-sized types will still be laid out "in-place" in storage, even if they contain dynamically-sized types (i.e. if they are dynamically-encoded - in calldata that means they will themselves use tail pointers, but not in storage).

And in general we should take care that we don't zero out any slot twice.

So thinking about this again, one option to do this would be something like this:

```
    std::set<u256> slotsCleared;
  	for (auto member: structMembers)
		if(member.type->storageSize < 32)
        {
          auto const& slot = structType.storageOffsetsOfMember(member.name).first;
          if (!slotsCleared.count(slot))
          {
            memberSetValues.emplace_back(...); // basically do ``sstore(slot, 0)``
            slotsCleared.emplace(slot);
          }
        }
        else
          memberSetValues.emplace_back(...); // use <setZero> on the member
```

At least I think so - but we need to be rather careful about getting this right :-).
On the other hand I just read people complaining about old-codegen being wasteful in this regard in https://gitter.im/ethereum/solidity, so it's probably worth the effort :-).

Reply:
(had to edit this a few times to arrive at something hopefully understandable)

Reply:
This will result basically in an unrolled loop setting most slots to zero, except for types that occupy more than 32 bytes in storage anyways. For those "at least one full slot"-types, the "setToZeroFunction" should either (for value types) just boil down to one or more ``sstore(slot, 0)`` as well, or (for dynamically sized reference types) follow the pointer and do a "deep clear".

Reply:
I think I did it as you explained here. I also think that this implementation now has minimum number of `sstore`s.
Also would be nice to compare gas usage with old codegen :)